crates/red_knot_python_semantic/src/program.rs

```diff
@@ -1,20 +1,52 @@
-use crate::Db;
-use ruff_db::system::SystemPathBuf;
+use anyhow::Context;
 use salsa::Durability;
+use salsa::Setter;
+
+use ruff_db::system::SystemPathBuf;
+
+use crate::module_resolver::SearchPaths;
+use crate::Db;
 
 #[salsa::input(singleton)]
 pub struct Program {
     pub target_version: TargetVersion,
 
+    #[default]
     #[return_ref]
-    pub search_paths: SearchPathSettings,
+    pub(crate) search_paths: SearchPaths,
 }
 
 impl Program {
-    pub fn from_settings(db: &dyn Db, settings: ProgramSettings) -> Self {
-        Program::builder(settings.target_version, settings.search_paths)
+    pub fn from_settings(db: &dyn Db, settings: ProgramSettings) -> anyhow::Result<Self> {
```

Comment:
Couldn't we return `Result<Self, SearchPathvalidationError>` here? That error type was originally public. I made it private when merging the two crates to avoid some clippy complaints, but it would be fine to make it public again IMO

Reply:
`from_settings` could potentially return more errors than just `SearchPathvalidationError`. For example, we could error if we don't support the target version. That's why I used `anyhow` (and we have no upstream code that matches on the error anyway).

Reply:
Currently it's impossible to have a target version here that we don't support because the `TargetVersion` enumeration only includes versions that we support, so clap's command-line validation should take care of that for us. But I see your point, thanks.